source/extensions/filters/http/abac/abac_filter.cc

```diff
@@ -0,0 +1,171 @@
+    }
+
+    auto ssl_info = info_.downstreamSslConnection();
+    if (ssl_info != nullptr) {
+      if (value == kCertificatePresented) {
+        return CelValue::CreateBool(ssl_info->peerCertificatePresented());
+      } else if (value == kSubjectLocalCertificate) {
+        return CelValue::CreateString(ssl_info->subjectLocalCertificate());
+      } else if (value == kIssuerPeerCertificate) {
+        return CelValue::CreateString(ssl_info->issuerPeerCertificate());
+      } else if (value == kSubjectPeerCertificate) {
+        return CelValue::CreateString(ssl_info->subjectPeerCertificate());
+      } else if (value == kTlsVersion) {
+        return CelValue::CreateString(ssl_info->tlsVersion());
+      }
+    }
+    return {};
+  }
+  int size() const override { return 0; }
+  bool empty() const override { return false; }
+  const CelList* ListKeys() const override { return nullptr; }
+
+private:
+  const StreamInfo::StreamInfo& info_;
+};
+
+class HeadersWrapper : public CelMap {
+public:
+  HeadersWrapper(const Http::HeaderMap& headers) : headers_(headers) {}
+
+  absl::optional<CelValue> operator[](CelValue key) const override {
+    if (!key.IsString()) {
+      return {};
+    }
+    auto out = headers_.get(Http::LowerCaseString(std::string(key.StringOrDie().value())));
+    if (out == nullptr) {
+      return {};
+    }
+    return CelValue::CreateString(out->value().getStringView());
+  }
+  int size() const override { return headers_.size(); }
+  bool empty() const override { return headers_.empty(); }
+  const CelList* ListKeys() const override { return nullptr; }
+  const Http::HeaderMap& headers_;
+};
+
+class RequestWrapper : public CelMap {
+public:
+  RequestWrapper(const Http::HeaderMap& headers, const StreamInfo::StreamInfo& info)
+      : wrapper_(headers), info_(info) {}
+
+  absl::optional<CelValue> operator[](CelValue key) const override {
+    if (!key.IsString()) {
+      return {};
+    }
+    auto value = key.StringOrDie().value();
+    if (value == kPath) {
+      return CelValue::CreateString(wrapper_.headers_.Path()->value().getStringView());
+    } else if (value == kHost) {
+      return CelValue::CreateString(wrapper_.headers_.Host()->value().getStringView());
+    } else if (value == kScheme) {
+      return CelValue::CreateString(wrapper_.headers_.Scheme()->value().getStringView());
+    } else if (value == kMethod) {
+      return CelValue::CreateString(wrapper_.headers_.Method()->value().getStringView());
+    } else if (value == kReferer) {
+      return CelValue::CreateString(wrapper_.headers_.Referer()->value().getStringView());
+    } else if (value == kHeaders) {
+      return CelValue::CreateMap(&wrapper_);
+    } else if (value == kTime) {
+      return CelValue::CreateTimestamp(absl::FromChrono(info_.startTime()));
+    }
+    return {};
+  }
+
+  int size() const override { return 0; }
+  bool empty() const override { return false; }
+  const CelList* ListKeys() const override { return nullptr; }
```

Comment:
Why are all listkeys returning null?

Reply:
Not implemented. For wrappers, it's undesired. For headers, it might be useful for comprehension operations (`all` or combining all header values), which I suppose is not common.